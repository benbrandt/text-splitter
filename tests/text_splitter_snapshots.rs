use std::fs;

use more_asserts::assert_le;
use once_cell::sync::Lazy;
use text_splitter::{Characters, ChunkSizer, TextSplitter};
use tiktoken_rs::{cl100k_base, CoreBPE};
use tokenizers::Tokenizer;

#[test]
fn characters_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default();
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(Characters.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn characters_range() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        let splitter = TextSplitter::default();
        let chunks = splitter.chunks(&text, 500..=2000).collect::<Vec<_>>();

        assert_eq!(chunks.join(""), text);
        for chunk in chunks.iter() {
            assert_le!(Characters.chunk_size(chunk), 2000);
        }
        insta::assert_yaml_snapshot!(chunks);
    });
}

#[test]
fn characters_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::default().with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static HUGGINGFACE_TOKENIZER: Lazy<Tokenizer> =
    Lazy::new(|| Tokenizer::from_pretrained("bert-base-cased", None).unwrap());

#[test]
fn huggingface_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(HUGGINGFACE_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn huggingface_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(HUGGINGFACE_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

static TIKTOKEN_TOKENIZER: Lazy<CoreBPE> = Lazy::new(|| cl100k_base().unwrap());

#[test]
fn tiktoken_default() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone());
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            assert_eq!(chunks.join(""), text);
            for chunk in chunks.iter() {
                assert_le!(TIKTOKEN_TOKENIZER.chunk_size(chunk), chunk_size);
            }
            insta::assert_yaml_snapshot!(chunks);
        }
    });
}

#[test]
fn tiktoken_trim() {
    insta::glob!("inputs/text/*.txt", |path| {
        let text = fs::read_to_string(path).unwrap();

        for chunk_size in [10, 100, 1000] {
            let splitter = TextSplitter::new(TIKTOKEN_TOKENIZER.clone()).with_trim_chunks(true);
            let chunks = splitter.chunks(&text, chunk_size).collect::<Vec<_>>();

            insta::assert_yaml_snapshot!(chunks);
        }
    });
}
