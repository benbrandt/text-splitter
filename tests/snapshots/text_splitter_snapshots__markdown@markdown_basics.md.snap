---
source: tests/text_splitter_snapshots.rs
expression: chunks
input_file: tests/inputs/markdown/markdown_basics.md
---
- "Markdown: "
- "Basics\n"
- "=========="
- "======\n\n"
- "<ul id=\""
- ProjectSub
- "menu\">\n"
- "    <li><a"
- " href=\"/"
- projects/
- "markdown/\""
- " title=\""
- "Markdown "
- "Project "
- "Page\">Main"
- "</a></li>\n"
- "    <li><a"
- " class=\""
- "selected\" "
- "title=\""
- "Markdown "
- "Basics\">"
- Basics</a>
- "</li>\n"
- "    <li><a"
- " href=\"/"
- projects/
- markdown/
- "syntax\" "
- "title=\""
- "Markdown "
- "Syntax "
- Documentat
- "ion\">"
- Syntax</a>
- "</li>\n"
- "    <li><a"
- " href=\"/"
- projects/
- markdown/
- "license\" "
- "title=\""
- "Pricing "
- "and "
- "License "
- Informatio
- "n\">License"
- "</a></li>\n"
- "    <li><a"
- " href=\"/"
- projects/
- markdown/
- "dingus\" "
- "title=\""
- "Online "
- "Markdown "
- "Web Form\">"
- Dingus</a>
- "</li>\n"
- "</ul>\n"
- "\n\n"
- "Getting "
- "the Gist "
- "of "
- "Markdown's"
- " "
- Formatting
- " Syntax\n"
- "----------"
- "----------"
- "----------"
- "----------"
- "--------\n\n"
- "This page "
- "offers a "
- "brief "
- "overview "
- "of what "
- "it's like "
- "to use "
- "Markdown.\n"
- "The ["
- "syntax "
- "page] [s] "
- "provides "
- "complete, "
- "detailed "
- documentat
- "ion for\n"
- "every "
- "feature, "
- "but "
- "Markdown "
- "should be "
- "very easy "
- to pick up
- " simply by"
- "\n"
- looking at
- " a few "
- "examples "
- "of it in "
- "action. "
- "The "
- "examples "
- "on this "
- "page\n"
- "are "
- written in
- " a before/"
- "after "
- "style, "
- "showing "
- "example "
- syntax and
- " the\n"
- "HTML "
- "output "
- "produced "
- "by "
- "Markdown.\n"
- "\nIt'"
- "s also "
- helpful to
- " simply "
- "try "
- "Markdown "
- "out; the ["
- "Dingus] [d"
- "] is a\n"
- "web "
- applicatio
- "n that "
- allows you
- " type your"
- " own "
- Markdown-
- "formatted "
- "text\n"
- "and "
- "translate "
- "it to "
- "XHTML.\n\n"
- "**Note:** "
- "This "
- "document "
- "is itself "
- "written "
- "using "
- "Markdown; "
- "you\ncan ["
- "see the "
- source for
- " it by "
- "adding '."
- "text' to "
- "the URL] ["
- "src]."
- "\n\n"
- "  [s]: /"
- projects/
- markdown/
- "syntax  \""
- "Markdown "
- "Syntax\"\n"
- "  [d]: /"
- projects/
- markdown/
- "dingus  \""
- "Markdown "
- "Dingus\"\n"
- "  [src]: /"
- projects/
- markdown/
- basics.tex
- "t\n\n\n## "
- Paragraphs
- ", Headers,"
- " "
- Blockquote
- "s ##\n\n"
- "A "
- "paragraph "
- "is simply "
- "one or "
- "more "
- consecutiv
- e lines of
- " text, "
- "separated\n"
- "by one or "
- more blank
- " lines. "
- "(A blank "
- "line is "
- "any line "
- that looks
- " like\n"
- "a blank "
- "line -- a "
- "line "
- containing
- " nothing "
- but spaces
- " or tabs "
- "is\n"
- considered
- " blank.) "
- "Normal "
- paragraphs
- " should "
- "not be "
- "indented "
- "with\n"
- "spaces or "
- "tabs.\n\n"
- "Markdown "
- offers two
- " styles of"
- " headers: "
- "*Setext* "
- and *atx*.
- "\n"
- Setext-
- "style "
- "headers "
- "for `<h1>`"
- " and `<h2>"
- "` are "
- created by
- "\n\""
- underlinin
- "g\" with "
- "equal "
- "signs (`=`"
- ") and "
- "hyphens (`"
- "-`), "
- respective
- "ly.\n"
- "To create "
- an atx-
- "style "
- "header, "
- you put 1-
- "6 hash "
- "marks (`#`"
- ") at the\n"
- "beginning "
- "of the "
- "line -- "
- the number
- " of hashes"
- " equals "
- "the "
- "resulting\n"
- "HTML "
- "header "
- "level.\n\n"
- Blockquote
- "s are "
- "indicated "
- "using "
- email-
- "style '`>`"
- "' angle "
- "brackets.\n"
- "\nMarkdown:"
- "\n\n    "
- "A First "
- "Level "
- "Header\n"
- "    ======"
- "=========="
- "====\n\n    "
- "A Second "
- "Level "
- "Header\n"
- "    ------"
- "----------"
- "-----\n\n"
- "    Now is"
- " the time "
- "for all "
- "good men "
- to come to
- "\n    "
- the aid of
- " their "
- "country. "
- "This is "
- "just a\n"
- "    "
- "regular "
- paragraph.
- "\n\n    "
- "The quick "
- "brown fox "
- "jumped "
- "over the "
- "lazy\n    "
- "dog's back"
- ".\n\n    "
- "### Header"
- " 3\n\n    "
- "> This is "
- "a "
- blockquote
- ".\n    >\n"
- "    > This"
- " is the "
- "second "
- "paragraph "
- "in the "
- blockquote
- ".\n    >\n"
- "    > ## "
- This is an
- " H2 in a "
- blockquote
- "\n\n\nOutput:"
- "\n\n    "
- "<h1>A "
- "First "
- "Level "
- Header</h1
- ">\n\n    "
- "<h2>A "
- "Second "
- "Level "
- Header</h2
- ">\n\n    "
- "<p>Now is "
- "the time "
- "for all "
- "good men "
- to come to
- "\n    "
- the aid of
- " their "
- "country. "
- "This is "
- "just a\n"
- "    "
- "regular "
- paragraph.
- "</p>\n\n    "
- "<p>The "
- "quick "
- "brown fox "
- "jumped "
- "over the "
- "lazy\n    "
- "dog's back"
- ".</p>\n\n"
- "    <h3>"
- Header 3</
- "h3>\n\n    "
- "<"
- blockquote
- ">\n    "
- "    <p>"
- "This is a "
- blockquote
- ".</p>\n\n"
- "        <p"
- ">This is "
- the second
- " paragraph"
- " in the "
- blockquote
- ".</p>\n\n"
- "        <"
- h2>This is
- " an H2 in "
- "a "
- blockquote
- "</h2>\n    "
- "</"
- blockquote
- ">\n\n\n\n### "
- "Phrase "
- "Emphasis #"
- "##\n\n"
- "Markdown "
- "uses "
- "asterisks "
- "and "
- underscore
- "s to "
- "indicate "
- "spans of "
- "emphasis.\n"
- "\nMarkdown:"
- "\n\n    "
- "Some of "
- "these "
- words *are
- " "
- emphasized
- "*.\n    "
- "Some of "
- "these "
- words _are
- " "
- emphasized
- " also_.\n\n"
- "    Use "
- "two "
- "asterisks "
- for **
- "strong "
- emphasis**
- ".\n    "
- "Or, if you"
- " prefer, "
- "__use two "
- underscore
- "s "
- instead__.
- "\n\nOutput:"
- "\n\n    "
- "<p>Some of"
- " these "
- words <em>
- "are "
- emphasized
- "</em>.\n"
- "    Some "
- "of these "
- words <em>
- "are "
- emphasized
- " also</em>"
- ".</p>\n\n"
- "    <p>Use"
- " two "
- "asterisks "
- for <
- strong>
- "strong "
- emphasis</
- "strong>.\n"
- "    Or, if"
- " you "
- "prefer, <"
- strong>use
- " two "
- underscore
- s instead<
- /strong>.<
- "/p>\n\n\n\n## "
- "Lists ##\n\n"
- "Unordered "
- (bulleted)
- " lists use"
- " asterisks"
- ", pluses, "
- "and "
- "hyphens (`"
- "*`,\n`+`"
- ", and `-`)"
- " as list "
- "markers. "
- "These "
- "three "
- "markers "
- "are\n"
- interchang
- able; this
- ":\n\n    "
- "*   Candy."
- "\n    "
- "*   Gum.\n"
- "    *   "
- "Booze.\n\n"
- "this:"
- "\n\n    "
- +   Candy.
- "\n    "
- "+   Gum.\n"
- "    +   "
- "Booze.\n\n"
- "and this:"
- "\n\n    "
- "-   Candy."
- "\n    "
- "-   Gum.\n"
- "    -   "
- "Booze.\n\n"
- "all "
- "produce "
- "the same "
- "output:\n\n"
- "    <ul>\n"
- "    <li>"
- Candy.</li
- ">\n    "
- "<li>Gum.</"
- "li>\n    "
- "<li>Booze."
- "</li>\n    "
- "</ul>\n\n"
- Ordered (
- "numbered) "
- "lists use "
- "regular "
- "numbers, "
- "followed "
- by periods
- ", as\n"
- "list "
- "markers:\n\n"
- "    "
- "1.  Red\n"
- "    "
- "2.  Green\n"
- "    "
- "3.  Blue\n\n"
- "Output:"
- "\n\n    "
- "<ol>\n    "
- "<li>Red</"
- "li>\n    "
- "<li>Green<"
- "/li>\n    "
- "<li>Blue</"
- "li>\n    "
- "</ol>\n\n"
- If you put
- " blank "
- "lines "
- "between "
- "items, "
- "you'll get"
- " `<p>` "
- "tags for "
- "the\n"
- "list item "
- "text. "
- "You can "
- "create "
- multi-
- "paragraph "
- list items
- " by "
- "indenting\n"
- "the "
- paragraphs
- " by 4 "
- "spaces or "
- "1 tab:\n\n"
- "    *   A "
- list item.
- "\n\n    "
- "    With "
- "multiple "
- paragraphs
- ".\n\n    "
- "*   "
- "Another "
- "item in "
- "the list.\n"
- "\nOutput:"
- "\n\n    "
- "<ul>\n    "
- "<li><p>A "
- list item.
- "</p>\n    "
- "<p>With "
- "multiple "
- paragraphs
- ".</p></li>"
- "\n    "
- "<li><p>"
- "Another "
- "item in "
- the list.<
- "/p></li>\n"
- "    </ul>\n"
- "\n\n\n### "
- Links
- " ###\n\n"
- "Markdown "
- "supports "
- two styles
- " for "
- "creating "
- "links: *"
- "inline* "
- "and\n"
- "*reference"
- "*"
- ". "
- "With both "
- "styles, "
- "you use "
- "square "
- "brackets "
- to delimit
- " the\n"
- "text you "
- "want to "
- "turn into "
- "a link.\n\n"
- Inline-
- "style "
- "links use "
- parenthese
- "s "
- immediatel
- "y after "
- "the link "
- "text.\n"
- "For "
- "example:\n\n"
- "    This "
- "is an ["
- "example "
- "link](http"
- "://"
- example.co
- "m/).\n\n"
- "Output:"
- "\n\n    "
- "<p>This is"
- " an <a "
- "href=\"http"
- "://"
- example.co
- "m/\">\n    "
- "example "
- link</a>.<
- "/p>\n\n"
- Optionally
- ", you may "
- "include a "
- "title "
- "attribute "
- "in the "
- parenthese
- "s:\n\n    "
- This is an
- " [example "
- "link](http"
- "://"
- example.co
- "m/ \"With a"
- " Title\").\n"
- "\nOutput:"
- "\n\n    "
- "<p>This is"
- " an <a "
- "href=\"http"
- "://"
- example.co
- "m/\" title="
- "\"With a "
- "Title\">\n"
- "    "
- "example "
- link</a>.<
- "/p>\n\n"
- Reference-
- "style "
- "links "
- "allow you "
- "to refer "
- "to your "
- "links by "
- "names, "
- "which\n"
- you define
- " elsewhere"
- " in your "
- "document:\n"
- "\n    "
- "I get 10 "
- times more
- " traffic "
- "from ["
- "Google][1]"
- " than from"
- "\n    "
- "[Yahoo][2]"
- " or [MSN]["
- "3].\n\n    "
- "[1]: http:"
- //
- google.com
- "/        \""
- "Google\"\n"
- "    [2]: "
- "http://"
- search.yah
- "oo.com/  \""
- "Yahoo "
- "Search\"\n"
- "    [3]: "
- "http://"
- search.msn
- ".com/    \""
- MSN Search
- "\"\n\nOutput:"
- "\n\n    "
- "<p>I get "
- "10 times "
- "more "
- "traffic "
- "from <a "
- "href=\"http"
- "://"
- google.com
- "/\"\n    "
- "title=\""
- "Google\">"
- Google</a>
- " than from"
- " <a href=\""
- "http://"
- search.yah
- "oo.com/\"\n"
- "    title="
- "\"Yahoo "
- "Search\">"
- "Yahoo</a> "
- or <a href
- "=\"http://"
- search.msn
- ".com/\"\n"
- "    title="
- "\"MSN "
- "Search\">"
- MSN</a>.</
- "p>\n\n"
- "The title "
- "attribute "
- "is "
- "optional. "
- Link names
- " may "
- "contain "
- "letters,\n"
- "numbers "
- and spaces
- ", but are "
- "*not* case"
- " sensitive"
- ":\n\n    "
- I start my
- " morning "
- with a cup
- " of coffee"
- " and\n    "
- "[The New "
- York Times
- "][NY Times"
- "].\n\n    "
- "[ny times]"
- ": http://"
- www.nytime
- "s.com/\n\n"
- "Output:"
- "\n\n    "
- "<p>I start"
- " my "
- "morning "
- with a cup
- " of coffee"
- " and\n    "
- "<a href=\""
- "http://"
- www.nytime
- "s.com/\">"
- "The New "
- York Times
- "</a>.</p>\n"
- "\n\n### "
- "Images ###"
- "\n\n"
- "Image "
- "syntax is "
- "very much "
- "like link "
- "syntax.\n\n"
- Inline (
- titles are
- " optional)"
- ":\n\n    "
- "!["
- "alt text]("
- /path/to/
- "img.jpg \""
- "Title\")\n\n"
- Reference-
- "style:\n\n"
- "    "
- "!["
- "alt text]["
- "id]\n\n    "
- "[id]: /"
- path/to/
- "img.jpg \""
- "Title\"\n\n"
- "Both of "
- "the above "
- "examples "
- "produce "
- "the same "
- "output:\n\n"
- "    <img "
- "src=\"/path"
- /to/
- "img.jpg\" "
- "alt=\"alt "
- "text\" "
- "title=\""
- "Title\" />\n"
- "\n\n\n### "
- "Code ###\n\n"
- "In a "
- "regular "
- "paragraph,"
- " you can "
- "create "
- "code span "
- "by "
- "wrapping "
- "text in\n"
- "backtick "
- "quotes. "
- "Any "
- ampersands
- " (`&`) and"
- " angle "
- brackets (
- "`<` or\n`>`"
- ") will "
- automatica
- "lly be "
- translated
- " into HTML"
- " entities."
- " This "
- "makes\n"
- it easy to
- " use "
- "Markdown "
- "to write "
- about HTML
- " example "
- "code:\n\n"
- "    I "
- "strongly "
- "recommend "
- "against "
- "using any "
- "`<blink>` "
- "tags.\n\n"
- "    I wish"
- " "
- SmartyPant
- "s used "
- "named "
- "entities "
- "like `&"
- "mdash;`\n"
- "    "
- instead of
- " decimal-"
- "encoded "
- "entities "
- "like `&#"
- "8212;`.\n\n"
- "Output:"
- "\n\n    "
- "<p>I "
- "strongly "
- "recommend "
- "against "
- "using any\n"
- "    <code>"
- "&lt;blink&"
- gt;</code>
- " tags.</p>"
- "\n\n    "
- "<p>I wish "
- SmartyPant
- "s used "
- "named "
- "entities "
- "like\n    "
- "<code>&amp"
- ;mdash;</
- "code> "
- instead of
- " decimal-"
- "encoded\n"
- "    "
- "entities "
- like <code
- ">&amp;#"
- 8212;</
- code>.</p>
- "\n\n\n"
- To specify
- " an entire"
- " block of "
- pre-
- "formatted "
- "code, "
- "indent "
- every line
- " of\n"
- "the block "
- "by 4 "
- "spaces or "
- "1 tab. "
- "Just like "
- "with code "
- "spans, `&`"
- ", `<`"
- ",\nand `>`"
- " "
- characters
- " will be "
- "escaped "
- automatica
- "lly.\n\n"
- "Markdown:"
- "\n\n    "
- "If you "
- "want your "
- "page to "
- "validate "
- "under "
- "XHTML 1.0 "
- "Strict,\n"
- "    you've"
- " got to "
- "put "
- "paragraph "
- "tags in "
- "your "
- blockquote
- "s:\n\n    "
- "    <"
- blockquote
- ">\n    "
- "        <p"
- ">For "
- example.</
- "p>\n    "
- "    </"
- blockquote
- ">\n\nOutput:"
- "\n\n    "
- "<p>If you "
- "want your "
- "page to "
- "validate "
- "under "
- "XHTML 1.0 "
- "Strict,\n"
- "    you've"
- " got to "
- "put "
- "paragraph "
- "tags in "
- "your "
- blockquote
- "s:</p>\n\n"
- "    <pre><"
- code>&lt;
- blockquote
- "&gt;\n    "
- "    &lt;p&"
- "gt;For "
- example.&
- "lt;/p&gt;\n"
- "    &lt;/"
- blockquote
- "&gt;\n    "
- "</code></"
- "pre>\n"

